ui/components/component-library/text/text.tsx

```diff
@@ -1,84 +1,76 @@
+      return 'span';
     // TextVariant.bodyLgMedium, TextVariant.bodyMd, TextVariant.bodyMdBold, TextVariant.bodySm, TextVariant.bodySmBold, TextVariant.bodyXs use default 'p' tag
     default:
-      return ValidTag.P;
+      return 'p';
   }
 };
 
-export const Text = forwardRef(function Text(
-  {
-    variant = TextVariant.bodyMd,
-    color = TextColor.textDefault,
-    fontWeight,
-    fontStyle,
-    textTransform,
-    textAlign,
-    textDirection,
-    overflowWrap,
-    ellipsis,
-    as,
-    className = '',
-    children,
-    ...props
-  }: TextProps,
-  ref: Ref<HTMLElement>,
-) {
-  // Check if as is set otherwise set a default tag based on variant
-  const Tag = as ?? getTextElementDefault(variant);
-  let strongTagFontWeight;
-
-  if (Tag === 'strong') {
-    strongTagFontWeight = FontWeight.Bold;
-  }
-
-  const computedClassName = classnames(
-    'mm-text',
-    className,
-    `mm-text--${variant}`,
+export const Text: TextComponent = React.forwardRef(
+  <C extends React.ElementType = 'div'>(
     {
-      [`mm-text--font-weight-${strongTagFontWeight || fontWeight}`]: Boolean(
-        strongTagFontWeight || fontWeight,
-      ),
-      [`mm-text--font-style-${String(fontStyle)}`]: Boolean(fontStyle),
-      [`mm-text--ellipsis`]: Boolean(ellipsis),
-      [`mm-text--text-transform-${String(textTransform)}`]:
-        Boolean(textTransform),
-      [`mm-text--text-align-${String(textAlign)}`]: Boolean(textAlign),
-      [`mm-text--overflow-wrap-${String(overflowWrap)}`]: Boolean(overflowWrap),
-    },
-  );
-
-  return (
-    <Box
-      className={classnames(computedClassName)}
-      as={Tag}
-      dir={textDirection}
-      color={color}
-      ref={ref}
-      {...props}
-    >
-      {children}
-    </Box>
-  );
-});
+      variant = TextVariant.bodyMd,
+      fontWeight,
+      fontStyle,
+      textTransform,
+      textAlign,
+      textDirection,
+      overflowWrap,
+      ellipsis,
+      as,
+      className = '',
+      children,
+      ...props
+    }: TextProps<C>,
+    ref?: PolymorphicRef<C>,
+  ) => {
+    // Set tag based on variant
+    // If as prop is passed tag will be overridden
+    const tag = getTextElementDefault(variant);
+    const computedClassName = classnames(
+      'mm-text',
+      className,
+      `mm-text--${variant}`,
+      {
+        [`mm-text--font-weight-${fontWeight}`]: Boolean(fontWeight),
+        [`mm-text--font-style-${fontStyle}`]: Boolean(fontStyle),
+        [`mm-text--ellipsis`]: Boolean(ellipsis),
+        [`mm-text--text-transform-${textTransform}`]: Boolean(textTransform),
+        [`mm-text--text-align-${textAlign}`]: Boolean(textAlign),
+        [`mm-text--overflow-wrap-${overflowWrap}`]: Boolean(overflowWrap),
+      },
+    );
+    return (
+      <Box
+        className={classnames(computedClassName)}
+        as={tag}
+        dir={textDirection}
+        ref={ref}
+        {...props}
+      >
+        {children}
+      </Box>
+    );
+  },
+);
```

Comment:
This shows how the pattern for polymorphic props would work in the `Text` component there is a little more overhead to just extending `BoxProps`